### PR TITLE
[sdk-logs+traces] Add internal processor weight api

### PR DESCRIPTION
**Pull request by CodeBlanch**
[Originally part of #5400]

## Changes

* Adds an internal pipeline weight concept to help order processors added to logging & tracing pipelines.

## Details

The design on #5400 calls for the OtlpExporter to be added at the end of the pipeline regardless of the call order. There isn't currently a way to do that so what this PR does is invent a way to make that work as an internal mechanism. This is more of a quick and dirty than a fully vetted solution and a lot of work/discussion will need to happen before something is exposed.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated

### CODE REVIEW

File: src/OpenTelemetry/ProcessorPipelineWeight.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    PipelineStart = int.MinValue,
+
+    /// <summary>
+    /// Pipeline enrichment. Value: <c>-10000</c>.
+    /// </summary>
+    /// <remarks>
+    /// Note: Enrichment processors which modify telemetry typically need to run
+    /// before exporters.
+    /// </remarks>
+    PipelineEnrichment = -10_000,
```
Comment (reyang) [inline src/OpenTelemetry/ProcessorPipelineWeight.cs:23]: Why is this needed?

File: src/OpenTelemetry/ProcessorPipelineWeight.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+/// <summary>
+/// Contains definitions for common processor pipeline weights.
+/// </summary>
+internal enum ProcessorPipelineWeight
+{
+    /// <summary>
+    /// Pipeline start. Value: <see cref="int.MinValue"/>.
+    /// </summary>
+    PipelineStart = int.MinValue,
```
Comment (reyang) [inline src/OpenTelemetry/ProcessorPipelineWeight.cs:14]: This is misleading - if there are multiple processors with this value, only one would be the start.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 3]
```diff
@@ -74,7 +72,11 @@ internal sealed class TracerProviderSdk : TracerProvider
             }
         }
 
-        foreach (var processor in state.Processors)
+        var processors = (IEnumerable<BaseProcessor<Activity>>)state.Processors.OrderBy(p => (int)p.PipelineWeight);
```
Comment (reyang) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:75]: Is this a stable sorting algorithm or not?

**Comment by reyang** [comment]
Couple feedback:
1. Scope it down - many things in the current PR are not required to unblock #5400.
2. I don't think this is the right direction - it is essentially introducing another DOS TSR / Windows Hook / Interrupt Handler chain https://www.ibm.com/docs/en/zvm/7.2?topic=interrupts-first-level-io-interrupt-processing.
3. I'm fine of having a workaround as long as a) everything is internal/private b) the scope is small enough and there is no extra piece which we don't need for now

File: src/OpenTelemetry/ProcessorPipelineWeight.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/ProcessorPipelineWeight.cs:14]: Ya, it is more like ranges. This is why I called it "weight" and not "order". My thinking was if a bunch of registrations happened something like this...

```csharp
builder.AddProcessor(new ProcessorA() { PipelineWeight = ProcessorPipelineWeight.PipelineEnrichment });
builder.AddProcessor(new ProcessorB() { PipelineWeight = ProcessorPipelineWeight.PipelineStart });
builder.AddProcessor(new ProcessorC() { PipelineWeight = ProcessorPipelineWeight.PipelineExporter });
builder.AddProcessor(new ProcessorD() { PipelineWeight = ProcessorPipelineWeight.PipelineEnrichment });
builder.AddProcessor(new ProcessorE() { PipelineWeight = ProcessorPipelineWeight.PipelineStart });
```

...would yield a pipeline weighted as `ProcessorB -> ProcessorE -> ProcessorA -> ProcessorD -> ProcessorC` which would probably work for most users.

You can get absolute control if you _want_ for example this...

```csharp
builder.AddProcessor(new ProcessorA() { PipelineWeight = ProcessorPipelineWeight.PipelineEnrichment + 1 });
builder.AddProcessor(new ProcessorB() { PipelineWeight = ProcessorPipelineWeight.PipelineStart + 1 });
builder.AddProcessor(new ProcessorC() { PipelineWeight = ProcessorPipelineWeight.PipelineExporter });
builder.AddProcessor(new ProcessorD() { PipelineWeight = ProcessorPipelineWeight.PipelineEnrichment });
builder.AddProcessor(new ProcessorE() { PipelineWeight = ProcessorPipelineWeight.PipelineStart });
```

...yields a pipeline weighted as `ProcessorE -> ProcessorB -> ProcessorD -> ProcessorA -> ProcessorC`.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs:75]: If we believe the [docs](https://learn.microsoft.com/dotnet/api/system.linq.enumerable.orderby), it is stable:

> This method performs a stable sort; that is, if the keys of two elements are equal, the order of the elements is preserved. In contrast, an unstable sort does not preserve the order of elements that have the same key.

File: src/OpenTelemetry/ProcessorPipelineWeight.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/ProcessorPipelineWeight.cs:23]: I'll remove. Some of this was me exploring how it might work for other cases.

**Comment by CodeBlanch** [comment]
@reyang Updated to remove the enum.

File: test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs [hunk 4]
```diff
@@ -134,8 +134,22 @@ public void LoggerProviderBuilderAddProcessorTest()
 
         using (var provider = Sdk.CreateLoggerProviderBuilder()
             .AddProcessor<CustomProcessor>()
-            .AddProcessor(sp => new CustomProcessor())
-            .AddProcessor(new CustomProcessor())
+            .AddProcessor(new CustomProcessor()
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs:137]: I understand this is internal, but we want to make sure that - the current behavior is retained as-is. i.e processors are called in the same order as they were added.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs:75]: > If we believe the docs

How should I interpret it? - "YES" or "I'm not sure"

I think @cijothomas asked the same question https://github.com/open-telemetry/opentelemetry-dotnet/pull/5424/files#r1517025574. The answer we expect is either "YES, it is stable and it will keep the existing behavior" or "NO, I realized there is something wrong and I'll update the PR to fix it".

**Review by reyang — APPROVED** [review]
LGTM once https://github.com/open-telemetry/opentelemetry-dotnet/pull/5424/files#r1516983320 is confirmed.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 3 above)
Comment (utpilla) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:75]: `PipelineWeight` is an `int`.

```suggestion
        var processors = (IEnumerable<BaseProcessor<Activity>>)state.Processors.OrderBy(p => p.PipelineWeight);
```

File: test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs [hunk 5]
```diff
@@ -154,11 +169,15 @@ public void LoggerProviderBuilderAddProcessorTest()
                 processors.Add(processor);
                 Assert.False(processor.Disposed);
 
+                Assert.True((int)processor.PipelineWeight >= lastWeight);
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs:172]: I think we could make this test more readable. Could we add a name field to `CustomProcessor`? We could then name the `CustomProcessor` instances and check if they were added in order. Something like add five instances of `CustomProcessor` : A, B, C, D, and E with different weights and expect them to be added in the right order such as B, C, A, D, E.




```

**Review by utpilla — APPROVED** [review]

File: test/OpenTelemetry.Tests/Trace/TracerProviderBuilderExtensionsTest.cs [hunk 6]
```diff
@@ -133,6 +133,75 @@ public void ServiceLifecycleAvailableToSDKBuilderTest()
...
+            var compositeProcessor = provider.Processor as CompositeProcessor<Activity>;
+
+            Assert.NotNull(compositeProcessor);
+
+            bool isFirstProcessor = true;
+            var lastWeight = int.MinValue;
+            var current = compositeProcessor.Head;
+            while (current != null)
+            {
+                if (isFirstProcessor)
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.Tests/Trace/TracerProviderBuilderExtensionsTest.cs:174]: nit: We could avoid condition here by keeping the first iteration out of while loop.

File: test/OpenTelemetry.Tests/Trace/TracerProviderBuilderExtensionsTest.cs [hunk 7]
```diff
@@ -133,6 +133,75 @@ public void ServiceLifecycleAvailableToSDKBuilderTest()
         Assert.True(myInstrumentation.Disposed);
     }
 
+    [Fact]
+    public void AddProcessorTest()
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.Tests/Trace/TracerProviderBuilderExtensionsTest.cs:137]: As there is a lot of discussion about not breaking the existing behavior, should we have one test with many processors, without PipelineWeight, to validate it?

**Review by rajkumar-rangaraj — APPROVED** [review]

**Review by vishweshbankwar — APPROVED** [review]

File: src/OpenTelemetry/Logs/LoggerProviderSdk.cs [hunk 8]
```diff
@@ -54,7 +54,7 @@ internal sealed class LoggerProviderSdk : LoggerProvider
         resourceBuilder.ServiceProvider = serviceProvider;
         this.Resource = resourceBuilder.Build();
 
-        foreach (var processor in state.Processors)
+        foreach (var processor in state.Processors.OrderBy(p => (int)p.PipelineWeight))
```
Comment (reyang) [inline src/OpenTelemetry/Logs/LoggerProviderSdk.cs:57]: Is "(int)" needed?

File: src/OpenTelemetry/BaseProcessor.cs [hunk 9]
```diff
@@ -27,6 +27,18 @@ public BaseProcessor()
...
+    /// Gets or sets the weight of the processor when added to the provider
+    /// pipeline. Default value: <c>0</c>.
+    /// </summary>
+    /// <remarks>
+    /// Note: Weight is used to order processors when building a provider
+    /// pipeline. Lower weighted processors come before higher weighted
+    /// processors. Changing the weight after a pipeline has been constructed
+    /// has no effect.
+    /// </remarks>
+    internal int PipelineWeight { get; set; }
```
Comment (reyang) [inline src/OpenTelemetry/BaseProcessor.cs:40]: I haven't spent time thinking about a proper name, but "pipeline weight" sounds wrong to me. Not a blocker though.

File: src/OpenTelemetry/ProcessorPipelineWeight.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry/ProcessorPipelineWeight.cs]: I guess I know where you're coming from, probably no need to spend much time on this discussion since I feel we're trying to solve a simple problem by introducing a complex series of issues. Happy to discuss with you offline when #5400 is done.

File: test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs (hunk 4 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs]: I don't think the tests as they were verified any order, only that the expected number of things were added. I updated to add more verification

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: It was really a question for you. Do you trust the docs? If yes, we're good. If no, I need to go and do more work 🤣

There's seems to be a test in place:

https://github.com/dotnet/runtime/blob/40ac297a1101ac293a1bae01f2e2e1556f7ec017/src/libraries/System.Linq/tests/OrderByTests.cs#L149-L172

But I also made our tests verify it: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5424/commits/62a9f40cdb55debe9a2aa6e42fe6c1f50e3faf67

File: test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs (hunk 5 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Logs/LoggerProviderBuilderExtensionsTests.cs]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5424/commits/62a9f40cdb55debe9a2aa6e42fe6c1f50e3faf67
